samples/nrf9160/serial_lte_modem/src/slm_at_tcpip.c

```diff
@@ -0,0 +1,716 @@
+	{AT_TCP_SEND, "AT#XTCPSEND", "at#xtcpsend",
+		handle_at_tcp_send},
+	{AT_TCP_RECV, "AT#XTCPRECV", "at#xtcprecv",
+		handle_at_tcp_recv},
+	{AT_UDP_SENDTO, "AT#XUDPSENDTO", "at#xudpsendto",
+		handle_at_udp_sendto},
+	{AT_UDP_RECVFROM, "AT#XUDPRECVFROM", "at#xudprecvfrom",
+		handle_at_udp_recvfrom},
+};
+
+static struct sockaddr_storage remote;
+
+static struct tcpip_client {
+	int sock; /* Socket descriptor. */
+	enum net_ip_protocol ip_proto; /* IP protocol */
+	bool connected; /* TCP connected flag */
+	slm_at_callback_t callback;
+} client;
+
+/* global variable defined in different files */
+extern struct at_param_list m_param_list;
+
+/**@brief Check whether a string has valid IPv4 address or not
+ */
+static bool check_for_ipv4(const char *address, u8_t length)
+{
+	int index;
+
+	for (index = 0; index < length; index++) {
+		char ch = *(address + index);
+		if ( (ch == '.') || (ch >= '0' && ch <= '9')) {
+			continue;
+		} else {
+			return false;
+		}
+	}
+
+	return true;
+}
+
+/**@brief Resolves host IPv4 address and port
+ */
+static int parse_host_by_ipv4(const char *ip, u16_t port)
+{
+	struct sockaddr_in *address4 = ((struct sockaddr_in *)&remote);
+
+	address4->sin_family = AF_INET;
+	address4->sin_port = htons(port);
+	LOG_DBG("IPv4 Address %s", log_strdup(ip));
+	/* NOTE inet_pton() returns 1 as success */
+	if (inet_pton(AF_INET, ip, &address4->sin_addr) == 1) {
+		return 0;
+	} else {
+		return -EINVAL;
+	}
+}
+
+
+static int parse_host_by_name(const char *name, u16_t port, int socktype)
+{
+	int err;
+	struct addrinfo *result;
+	struct addrinfo hints = {
+		.ai_family = AF_INET,
+		.ai_socktype = socktype
+	};
+	char ipv4_addr[NET_IPV4_ADDR_LEN];
+
+	err = getaddrinfo(name, NULL, &hints, &result);
+	if (err) {
+		LOG_ERR("ERROR: getaddrinfo failed %d", err);
+		return err;
+	}
+
+	if (result == NULL) {
+		LOG_ERR("ERROR: Address not found\n");
+		return -ENOENT;
+	}
+
+	/* IPv4 Address. */
+	struct sockaddr_in *server4 = ((struct sockaddr_in *)&remote);
+
+	server4->sin_addr.s_addr =
+		((struct sockaddr_in *)result->ai_addr)->sin_addr.s_addr;
+	server4->sin_family = AF_INET;
+	server4->sin_port = htons(port);
+
+	inet_ntop(AF_INET, &server4->sin_addr.s_addr, ipv4_addr,
+		  sizeof(ipv4_addr));
+	LOG_DBG("IPv4 Address found %s\n", ipv4_addr);
+
+	/* Free the address. */
+	freeaddrinfo(result);
+
+	return 0;
+}
+
+static int do_socket_open(u8_t type)
+{
+	int ret = 0;
+	char buf[16];
+
+	if (type == SOCK_STREAM) {
+		client.sock = socket(AF_INET, SOCK_STREAM,
+				IPPROTO_TCP);
+		client.ip_proto = IPPROTO_TCP;
+	} else if (type == SOCK_DGRAM) {
+		client.sock = socket(AF_INET, SOCK_DGRAM,
+				IPPROTO_UDP);
+		client.ip_proto = IPPROTO_UDP;
+	}
+	if (client.sock < 0) {
+		LOG_ERR("socket() failed: %d", -errno);
+		sprintf(buf, "#XSOCKET: %d\r\n", -errno);
+		client.callback(buf);
+		client.ip_proto = IPPROTO_IP;
+		ret = -errno;
+	} else {
+		sprintf(buf, "#XSOCKET: %d, %d\r\n", client.sock, client.ip_proto);
```

Comment:
This string is at minimum 16 bytes + null terminator, this will exceed the allocated `buf` size.

Reply:
Like the gps code, use a global 64-byte buffer now.